src/lando/main/scm/git.py

```diff
@@ -0,0 +1,298 @@
+from contextlib import contextmanager
+from pathlib import Path
+from typing import ContextManager, Optional
+
+from lando.main.scm.exceptions import SCMException
+
+from .abstract_scm import AbstractSCM
+
+logger = logging.getLogger(__name__)
+
+ENV_COMMITTER_NAME = "GIT_COMMITTER_NAME"
+ENV_COMMITTER_EMAIL = "GIT_COMMITTER_EMAIL"
+
+
+class GitSCM(AbstractSCM):
+    DEFAULT_ENV = {
+        "GIT_SSH_COMMAND": (
+            'ssh -o "StrictHostKeyChecking no" -o "PasswordAuthentication no"'
+        ),
+    }
+
+    """An implementation of the AbstractVCS for Git, for use by the Repo and LandingWorkers."""
+
+    default_branch: str
+
+    def __init__(self, path: str, default_branch: str = "main"):
+        self.default_branch = default_branch
+        super().__init__(path)
+
+    def clone(self, source):
+        """Clone a repository from a source.
+        Args:
+            source: The source to clone the repository from.
+        Returns:
+            None
+        """
+        self._git_run("clone", source, self.path)
+
+    def push(
+        self, push_path: str, target: Optional[str] = None, force_push: bool = False
+    ):
+        """Push local code to the remote repository.
+
+        Parameters:
+            push_path (str): The path to the repository where changes will be pushed.
+            target (Optional[str]): The target branch or reference to push to. Defaults to None.
+            force_push (bool): If True, force the push even if it results in a non-fast-forward update. Defaults to False.
+
+        Returns:
+            None
+        """
+        command = ["push"]
+        if force_push:
+            command += ["--force"]
+        command += [push_path]
+        if target:
+            command += [f"HEAD:{target}"]
+        self._git_run(*command)
+
+    @property
+    def REJECTS_PATH(self) -> Path:
+        return Path(self.path)
+
+    def last_commit_for_path(self, repo_path: str, path: str) -> str:
+        """Find last commit to touch a path.
+
+        Args:
+            repo_path (str): The path to the repository.
+            path (str): The specific path within the repository.
+
+        Returns:
+            str: The commit id
+        """
+        command = ["log", "--max-count=1", "--format=%H", "--", path]
+        result = self._git_run(*command, cwd=repo_path)
+        if not result:
+            raise SCMException(
+                "Empty data when determining the last commit touching a find; {repo_path=}",
+                result.stdout,
+                result.stderr,
+            )
+        return result
+
+    def apply_patch(
+        self, diff: str, commit_description: str, commit_author: str, commit_date: str
+    ):
+        """Apply the given patch to the current repository
+
+        Args:
+            patch_buf (StringIO): The patch to apply
+
+        Returns:
+            None
+        """
+        self._git_apply_patch(diff, commit_description, commit_author, commit_date)
```

Comment:
Since `apply_patch` just calls `_git_apply_patch` with the same function signature, we can remove `_git_apply_patch`.